### PR TITLE
Manifest V3 build

### PR DESCRIPTION
**Pull request by bershanskiy**
This PR adds a new build output, `build/*/chrome-mv3/`, which uses Manifest V3 format. This build provides basic functionality, but there are still challenges which need to be resolved:

 - Dark Reader's current script/style injection logic is not considered secure by Chrome's standards and therefore script/style injectionsfail
 - Service workers do not have access to `window` object and therefore Automation CSS media query fails
 - There is a terrible power use spike with this build, my laptop overheats when I install this build

### CODE REVIEW

**Comment by Gusted** [comment]
>     * Dark Reader's current script/style injection logic is not considered secure by Chrome's standards and therefore script/style injectionsfail

What are their recommended ways?

**Comment by bershanskiy** [comment]
Google recommends `chrome.scripting` for injecting code into extension's "clean context". But since we need to inect into page's own context, we probably need to use event handler method.

**Comment by Gusted** [comment]
> Google recommends `chrome.scripting` for injecting code into extension's "clean context".

I see, it's more or less the current chrome.tabs.executeScript but you are even limited more. Great :D The current script we are injecting does really need the page's context due to the prototype overriding. 

The style doesn't really matter, as long it works(and can very easily be edited) and overrides the style's CSS. But the script is a must to have the correct page's context.

**Comment by alexanderby** [comment]
If some API is missing we can ask in a user group or file an issue in Chromium bug tracker.

**Comment by Gusted** [comment]
> If some API is missing we can ask in a user group or file an issue in Chromium bug tracker.

That's possible, I don't have really good experience around their bug tracker nor the user groups 🤷🏽. But currently we just add a script into the page's HTML which is currently reliable(with #4977)

**Comment by bershanskiy** [comment]
> If some API is missing we can ask in a user group or file an issue in Chromium bug tracker.

We probably should. But injecting code into page's own context was never officially supported, it was always an afterthought.

File: tasks/bundle-js.js [hunk 1]
```diff
@@ -28,6 +30,19 @@ function patchFirefoxJS(/** @type {string} */code) {
...
 }
 
+function patchMV3JS(/** @type {string} */code) {
+    // MV3 moves a few APIs around
+    code = replace(code, 'chrome.browserAction.setIcon', 'chrome.action.setIcon');
+    code = replace(code, 'chrome.browserAction.setBadgeBackgroundColor', 'chrome.action.setBadgeBackgroundColor');
+    code = replace(code, 'chrome.browserAction.setBadgeText', 'chrome.action.setBadgeText');
+
+    // Background serivice workers do not have access to "foreground" APIs like chrome.i18n
+    code = replace(code, 'chrome.i18n.getUILanguage()', "navigator.language.split('-')[0]");
```
Comment (bershanskiy) [inline tasks/bundle-js.js:40]: Should I move this to `src/utils/locales.ts`?

Comment (alexanderby) [reply tasks/bundle-js.js:40]: Let's do this

**Comment by Gusted** [comment]
> We probably should. But injecting code into page's own context was never officially supported, it was always an afterthought.

https://groups.google.com/a/chromium.org/g/chromium-extensions/c/hQeJzPbG-js

**Comment by alexanderby** [comment]
Try applying this #4977 and see if it helps. I was scared to merge it, but let's see if it helps. If so, than let's merge.

**Comment by Gusted** [comment]
> Try applying this #4977 and see if it helps. I was scared to merge it, but let's see if it helps. If so, then let's merge.

It would enable the script in some places, but Manifest v3 is still another issue :)

File: tasks/bundle-js.js [hunk 2]
```diff
@@ -28,6 +30,19 @@ function patchFirefoxJS(/** @type {string} */code) {
...
+    // TODO: Fix me
+    code = replace(code, 'window.matchMedia(query).matches', 'false');
```
Comment (bershanskiy) [inline tasks/bundle-js.js:42]: There are multiple ways to implement CSS Media queries:
 - use content script context
 - spawn a "sandbox" which has an invisible window. This is the cleanest and the most reliable way, but spawning an entire process just to query a boolean value seems wasteful...

Comment (alexanderby) [reply tasks/bundle-js.js:42]: Then try doing it in context script

**Comment by alexanderby** [comment]
> > Try applying this #4977 and see if it helps. I was scared to merge it, but let's see if it helps. If so, then let's merge.
> 
> It would enable the script in some places, but Manifest v3 is still another issue :)

I was afraid it could disable it in other places as well :)

**Comment by Gusted** [comment]
> I was afraid it could disable it in other places as well :) Like Firefox.

Nope seems all good, `blob:` is special like normal it will match the `self` directive in CSP which sites like privacytools.io uses and some other places with strict CSP because in that context a blob made by dark reader is somewhat equal to a blob made on that site. It's a weird CSP but it luckily does the trick.

**Comment by alexanderby** [comment]
> > I was afraid it could disable it in other places as well :) Like Firefox.
> 
> Nope seems all good, `blob:` is special like normal it will match the `self` directive in CSP which sites like privacytools.io uses and some other places with strict CSP because in that context a blob made by dark reader is somewhat equal to a blob made on that site. It's a weird CSP but it luckily does the trick.

Well, then merge it.

**Comment by bershanskiy** [comment]
`blob:` is still considered insecure:

> Insecure CSP value "blob:" in directive 'script-src'.

**Comment by Gusted** [comment]
> `blob:` is still considered insecure:

`self` matches any blob: content, but any added scripts to the page's context relies on their CSP not on the extension's one.

**Comment by alexanderby** [comment]
I don't know, if it's not possible, then the fallbacks should work. We can file some issues. E.g. explain that the extensions need a way to watch for cssRules insertions and deletions. Or maybe allowing such scripts (though it seems insecure).

**Comment by Gusted** [comment]
> I don't know, if it's not possible, then the fallbacks should work. We can file some issues. E.g. explain that the extensions need a way to watch for cssRules insertions and deletions. Or maybe allowing such scripts (though it seems insecure).

The current fallback is something to avoid at all times(rAF polluting 60x times a second).

So far I've seen no chromium dev publicly stating(I'd might missed some conversiations) something about this whole manifest v3 mess and how to migrate for extensions with existing API's that will be gone. You might have luck with Firefox which seems to take a better stake at this whole manifest v3.

**Comment by alexanderby** [comment]
> The current fallback is something to avoid at all times(rAF polluting 60x times a second).

It just compares the count of cssRules. It should be fast, but not accurate.

> You might have luck with Firefox which seems to take a better stake at this whole manifest v3.

As far as I understand we can keep using MV2 there

**Comment by Gusted** [comment]
> It just compares the count of cssRules. It should be fast,

It is, until you will aggressively loop it 60x times It it's fast, but those little cpu time added together will become a problem over time, the main reason it was optimized away to have a passive listener as it seriously impacted performance of webpages and the profiler didn't like the function as well.

**Comment by alexanderby** [comment]
> Hmm, didn't they migrate to MV3 but keep comparability with MV2?

https://blog.mozilla.org/addons/2021/05/27/manifest-v3-update/
> We have not yet set a deprecation date for Manifest v2 but expect it to be supported for at least one year after Manifest v3 becomes stable in the release channel.

**Comment by Gusted** [comment]
> We have not yet set a deprecation date for Manifest v2 but expect it to be supported for at least one year after Manifest v3 becomes stable in the release channel.

That gives us some breathing time with firefox.

File: src/inject/dynamic-theme/index.ts [hunk 3]
```diff
@@ -134,13 +141,13 @@ function createStaticStyleOverrides() {
...
 
     const proxyScript = createOrUpdateScript('darkreader--proxy');
-    const blob = new Blob([`(${injectProxy})()`], {type: 'text/javascript'});
-    const url = URL.createObjectURL(blob);
-    proxyScript.src = url;
     proxyScript.textContent = '';
-    document.head.insertBefore(proxyScript, rootVarsStyle.nextSibling);
-    URL.revokeObjectURL(url);
-    proxyScript.remove();
+    proxyScript.addEventListener('yeet', injectedProxyEventListener);
```
Comment (Gusted) [inline src/inject/dynamic-theme/index.ts:145]: Could `yeet` be changed into `__darkreader__yeet`?

**Comment by bershanskiy** [comment]
This is ready for review, but we should merge #6564 bug fix prior to this.

File: tasks/bundle-js.js (hunk 2 above)
Comment (bershanskiy) [reply tasks/bundle-js.js:40]: Done!

**Review by alexanderby — APPROVED** [review]

**Pull request merged**